keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1709 @@
+- It will improve API composition. In particular CRDs that embed native types
+  (such as PodTemplate), which gain validation of the native type automatically.
+  This has the potential to simplify controller development and improve end
+  user experiences when using CRDs.
+- Validation is performed on versioned type, so error message and attributed field
+  path is far more likely to be relevant to the user.
+
+Please feel free to try out the
+[prototype](https://github.com/jpbetz/kubernetes/blob/cel-for-native/staging/src/k8s.io/sample-apiserver/pkg/apis/wardle/validation/README.md)
+to get hands on experience with this proposed enhancement.
+
+### Goals
+
+- Vast majority (95%+) of hand written validation are replaced with declarative
+validation. The remaining hand written rules will primarily be rules that _should
+not_ be published as part of the API, usually because server side state is
+involved in the validation decision.
+
+- `types.go` files become the de-facto IDL of Kubernetes for native types.
+It is worthing noting that `+enum` support, `+default` support and similar
+enhancements all moved our API development forward in this direction. This
+enhancement is an attempt to finish that story arc.
+
+- CRDs and native types are validated and published in OpenAPI in a consistent
+and uniform way. Improvement we make to CEL to enable declarative validation of
+native types will be made available to CRD authors and vis-versa.
+
+- No change to error message structure for clients. Difference to clients is 
+limited to minimal changes to error detail strings.
+
+### Non-Goals
+
+- It is not a goal for _all_ validation rules to be published to OpenAPI. There
+  exist complex, stateful rules that should not be published in OpenAPI. These
+  will remain "server side only" validation rules.
+- It is not a goal to require that _all_ validation rules be written in CEL. We
+  will continue to support hand written validation rules indefinitely for use
+  cases that are a poor fit using the existing validation mechanism (or
+  something very similar).
+
+
+## Proposal
+
+Declarative validation will be performed against versioned APIs. This differs from the hand
+written validation, which is evaluated against internal types.
+
+Go IDL tags will be added to support the following declarative validation rules:
+
+| Type of valiation      | Go IDL tag                                                       | OpenAPI validation field                                                           | Reason Type      |
+| ---------------------- | ---------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ---------------- |
+| string format          | `+format={format name}`                                          | `format`                                                                           | Invalid          |
```

Comment:
Is there a way that we could define our own custom k8s formats here?

Reply:
Yes, have the capabiltiy to add new formats. This is how we intend to handle most name formats for resources.